### PR TITLE
Added `CreateStruct` & Fixed `CreateMap`

### PR DESCRIPTION
**Pull request by cschuchardt88**
# Description

- Added `CreateStruct` method
- `CreateMap` changed to initialize items on the stack to be packed as a new map. Changed to follow `CreateArray` and `CreateStruct`.
- Type case removed because in C# `char` already is a `ushort`.

## Type of change

- [x] Optimization (the change is only an optimization)
- [x] Style (the change is only a code style for better maintenance or standard purpose)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

- [x] Unit tests

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by Jim8y** [comment]
can you add a little bit more explanation on this pr.

**Comment by roman-khimov** [comment]
Changes are nice, but I don't like the way they're stuffed into a single commit without any real comments (more efficient packer is not a "fix", additional method is an additional method, type cast is a separate matter as well).

**Comment by cschuchardt88** [comment]
I updated the description in the 1st comment.

File: src/Neo/VM/Helper.cs [hunk 1]
```diff
@@ -54,18 +54,64 @@ public static ScriptBuilder CreateArray<T>(this ScriptBuilder builder, IReadOnly
         /// <param name="builder">The <see cref="ScriptBuilder"/> to be used.</param>
         /// <param name="map">The key/value pairs of the map.</param>
         /// <returns>The same instance as <paramref name="builder"/>.</returns>
-        public static ScriptBuilder CreateMap<TKey, TValue>(this ScriptBuilder builder, IEnumerable<KeyValuePair<TKey, TValue>> map = null)
+        public static ScriptBuilder CreateMap<TKey, TValue>(this ScriptBuilder builder, IEnumerable<KeyValuePair<TKey, TValue>> map)
+            where TKey : notnull
+            where TValue : notnull
```
Comment (Jim8y) [inline src/Neo/VM/Helper.cs:59]: Value can not be null?

Comment (cschuchardt88) [reply src/Neo/VM/Helper.cs:59]: You can't do `map[null] = null'. You could before but wouldn't work. Needs to be `StackItem.Null`

**Review by shargon — APPROVED** [review]

**Review by Jim8y — APPROVED** [review]

File: src/Neo/VM/Helper.cs [hunk 2]
```diff
@@ -54,18 +54,64 @@ public static ScriptBuilder CreateArray<T>(this ScriptBuilder builder, IReadOnly
...
+            if (count == 0)
+                return builder.Emit(OpCode.NEWMAP);
+
+            foreach (var (key, value) in map.Reverse())
+            {
+                builder.EmitPush(value);
+                builder.EmitPush(key);
+            }
+            builder.EmitPush(count);
+            return builder.Emit(OpCode.PACKMAP);
```
Comment (AnnaShaleva) [inline src/Neo/VM/Helper.cs:72]: A really nice optimisation, it costs less GAS than the old code.

File: src/Neo/VM/Helper.cs [hunk 3]
```diff
@@ -218,7 +264,7 @@ public static ScriptBuilder EmitPush(this ScriptBuilder builder, object obj)
                     builder.EmitPush(data);
                     break;
                 case char data:
-                    builder.EmitPush((ushort)data);
+                    builder.EmitPush(data);
```
Comment (AnnaShaleva) [inline src/Neo/VM/Helper.cs:267]: Can we add a test for it?

**Review by AnnaShaleva — COMMENTED** [review]
Otherwise LGTM.

File: src/Neo/VM/Helper.cs [hunk 4]
```diff
@@ -54,18 +54,64 @@ public static ScriptBuilder CreateArray<T>(this ScriptBuilder builder, IReadOnly
...
-                    builder.EmitPush(p.Value);
-                    builder.Emit(OpCode.SETITEM);
-                }
-            return builder;
+            var count = map.Count();
+
+            if (count == 0)
+                return builder.Emit(OpCode.NEWMAP);
+
+            foreach (var (key, value) in map.Reverse())
```
Comment (shargon) [inline src/Neo/VM/Helper.cs:66]: Does we need the reverse?

Comment (cschuchardt88) [reply src/Neo/VM/Helper.cs:66]: yes

Comment (shargon) [reply src/Neo/VM/Helper.cs:66]: But it's more optimized if we change the tests, isn't it?

Comment (cschuchardt88) [reply src/Neo/VM/Helper.cs:66]: tests are fine

File: src/Neo/VM/Helper.cs (hunk 3 above)
Comment (cschuchardt88) [reply src/Neo/VM/Helper.cs:267]: https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/char

**Review by AnnaShaleva — APPROVED** [review]

File: src/Neo/VM/Helper.cs [hunk 5]
```diff
@@ -54,18 +54,64 @@ public static ScriptBuilder CreateArray<T>(this ScriptBuilder builder, IReadOnly
         /// <param name="builder">The <see cref="ScriptBuilder"/> to be used.</param>
         /// <param name="map">The key/value pairs of the map.</param>
         /// <returns>The same instance as <paramref name="builder"/>.</returns>
-        public static ScriptBuilder CreateMap<TKey, TValue>(this ScriptBuilder builder, IEnumerable<KeyValuePair<TKey, TValue>> map = null)
+        public static ScriptBuilder CreateMap<TKey, TValue>(this ScriptBuilder builder, IEnumerable<KeyValuePair<TKey, TValue>> map)
```
Comment (vncoelho) [inline src/Neo/VM/Helper.cs:57]: if the key and value are non-null before emitting the map structure. If previous code allowed null values, this new behavior could lead to unexpected failures where null handling was previously permissible.

**Review by vncoelho — COMMENTED** [review]
I am not sure if a hardfork is needed here. Perhaps needs @vang1ong7ang @dusmart

**Comment by roman-khimov** [comment]
It's a script builder, it doesn't affect byte code execution, so it's not hardfork-dependent.

File: src/Neo/VM/Helper.cs (hunk 4 above)
Comment (shargon) [reply src/Neo/VM/Helper.cs:66]: I know, but the test should be changed if fail, I think that it's faster to do it without this reverse

Comment (roman-khimov) [reply src/Neo/VM/Helper.cs:66]: I guess the expectation is that we have the same order (even though "order" doesn't make much sense for a map in general) of elements in the VM map as we have in a map passed to `CreateMap()`. Which means reversing elements either in VM (suboptimal) or here (much cheaper).

Comment (shargon) [reply src/Neo/VM/Helper.cs:66]: But we need to preserve the order in a map?

**Review by shargon — APPROVED** [review]

**Pull request merged**